pkg/webhooks/server.go

```diff
@@ -5,22 +5,110 @@ import (
+	// DumpOperations holds all operations that should be logged in debug mode.
+	DumpOperations []string
+	// DumpKinds holds all kinds that should be logged in debug mode.
+	DumpKinds []string
+}
+
+// AdmissionRequestPayload holds a copy of the AdmissionRequest payload
+type AdmissionRequestPayload struct {
+	UID                types.UID                    `json:"uid"`
+	Kind               metav1.GroupVersionKind      `json:"kind"`
+	Resource           metav1.GroupVersionResource  `json:"resource"`
+	SubResource        string                       `json:"subResource,omitempty"`
+	RequestKind        *metav1.GroupVersionKind     `json:"requestKind,omitempty"`
+	RequestResource    *metav1.GroupVersionResource `json:"requestResource,omitempty"`
+	RequestSubResource string                       `json:"requestSubResource,omitempty"`
+	Name               string                       `json:"name,omitempty"`
+	Namespace          string                       `json:"namespace,omitempty"`
+	Operation          string                       `json:"operation"`
+	UserInfo           authenticationv1.UserInfo    `json:"userInfo"`
+	Object             unstructured.Unstructured    `json:"object,omitempty"`
+	OldObject          unstructured.Unstructured    `json:"oldObject,omitempty"`
+	DryRun             *bool                        `json:"dryRun,omitempty"`
+	Options            unstructured.Unstructured    `json:"options,omitempty"`
+}
+
+func newAdmissionRequestPayload(rq *admissionv1.AdmissionRequest) (*AdmissionRequestPayload, error) {
+	newResource, oldResource, err := utils.ExtractResources(nil, rq)
+	if err != nil {
+		return nil, err
+	}
+	var options = new(unstructured.Unstructured)
+	if rq.Options.Raw != nil {
+		options, err = engineutils.ConvertToUnstructured(rq.Options.Raw)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return redactPayload(&AdmissionRequestPayload{
+		UID:                rq.UID,
+		Kind:               rq.Kind,
+		Resource:           rq.Resource,
+		SubResource:        rq.SubResource,
+		RequestKind:        rq.RequestKind,
+		RequestResource:    rq.RequestResource,
+		RequestSubResource: rq.RequestSubResource,
+		Name:               rq.Name,
+		Namespace:          rq.Namespace,
+		Operation:          string(rq.Operation),
+		UserInfo:           rq.UserInfo,
+		Object:             newResource,
+		OldObject:          oldResource,
+		DryRun:             rq.DryRun,
+		Options:            *options,
+	}), nil
+}
+
+func dumpAdmissionPayload(options *DebugModeOptions, request *admissionv1.AdmissionRequest) bool {
+	return options.DumpPayload && utils.SliceContains(options.DumpOperations, string(request.Operation)) &&
+		utils.SliceContains(options.DumpKinds, request.Kind.Kind)
+}
+
+func redactPayload(payload *AdmissionRequestPayload) *AdmissionRequestPayload {
+	if strings.EqualFold(payload.Kind.Kind, "Secret") {
+		if payload.Object.Object != nil {
+			unstructured.SetNestedField(payload.Object.Object, "**REDACTED**", "data")
```

Comment:
It's not iterating over the map, we need to redact values but keep keys in clear text.

Reply:
`{"apiVersion":"v1","data":"**REDACTED**","kind":"Secret","metadata":{"annotations":"**REDACTED**",`
do you mean the keys like `username, password` should be kept to look like this?
`{"apiVersion":"v1","data":{"username":"**REDACTED**", "password":"**REDACTED**"},"kind":"Secret","metadata":{"annotations":"**REDACTED**",`

Reply:
yes

Reply:
same thing for `annotations`